File: kafka-connect-rest-source/src/main/java/org/radarbase/connect/rest/request/RequestGenerator.java
/*
 * Copyright 2018 The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package org.radarbase.connect.rest.request;

import java.time.Instant;
import java.util.stream.Stream;
import org.apache.kafka.connect.storage.OffsetStorageReader;
import org.radarbase.connect.rest.config.RestSourceTool;

/**
 * Dynamically generates requests. The requests should be based on the offsets that are stored in
 * the response SourceRecord.
 */
public interface RequestGenerator extends RestSourceTool {
  Instant getTimeOfNextRequest();

  /**
   * Requests that should be queried next.
   */
  Stream<? extends RestRequest> requests();

  /**
   * Set the Kafka offset storage reader. This allows for resetting the request intervals.
   * @param offsetStorageReader possibly null offset storage reader.
   */
  void setOffsetStorageReader(OffsetStorageReader offsetStorageReader);
}

File: huawei-library/src/main/kotlin/org/radarbase/huawei/request/RequestGenerator.kt
/*
 * Copyright 2026 Onsentia
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package org.radarbase.huawei.request

import okhttp3.Response
import org.radarbase.huawei.converter.TopicData
import org.radarbase.huawei.route.Route
import org.radarbase.huawei.user.User

interface RequestGenerator {

    fun requests(user: User, max: Int): Sequence<RestRequest>

    fun requests(route: Route, user: User, max: Int): Sequence<RestRequest>

    fun requests(route: Route, max: Int): Sequence<RestRequest>

    fun requestSuccessful(request: RestRequest, response: Response): List<TopicData>

    fun requestFailed(request: RestRequest, response: Response): HuaweiError
}
